### PR TITLE
[core] Don't override external dashboard URL in internal KV store

### PR DESCRIPTION
**Pull request by nikitavemuri**
## Why are these changes needed?
Fix 2.0.0 release blocker bug where Ray State API and Jobs not accessible if the override URL doesn't support adding additional subpaths. This PR keeps the localhost dashboard URL in the internal KV store and only overrides in values printed or returned to the user.

Example output when `RAY_OVERRIDE_DASHBOARD_URL="https://external_dashboard_url/path1/?query_param1=val1&query_param2=val2"`:
- Create and connect to ray cluster
```
>> import ray
>> ray.init()
2022-08-15 23:05:45,275	WARNING worker.py:1284 -- File descriptor limit 2560 is too low for production servers and may result in connection errors. At least 8192 is recommended. --- Fix with 'ulimit -n 8192'
2022-08-15 23:05:47,821	INFO worker.py:1514 -- Started a local Ray instance. View the dashboard at https://external_dashboard_url/path1/?query_param1=val1&query_param2=val2.
Out[2]: RayContext(dashboard_url='external_dashboard_url/path1/?query_param1=val1&query_param2=val2', python_version='3.7.13', ray_version='3.0.0.dev0', ray_commit='{{RAY_COMMIT_SHA}}', address_info={'node_ip_address': '127.0.0.1', 'raylet_ip_address': '127.0.0.1', 'redis_address': None, 'object_store_address': '/tmp/ray/session_2022-08-15_23-05-45_286710_37962/sockets/plasma_store', 'raylet_socket_name': '/tmp/ray/session_2022-08-15_23-05-45_286710_37962/sockets/raylet', 'webui_url': 'external_dashboard_url/path1/?query_param1=val1&query_param2=val2', 'session_dir': '/tmp/ray/session_2022-08-15_23-05-45_286710_37962', 'metrics_export_port': 60950, 'gcs_address': '127.0.0.1:56494', 'address': '127.0.0.1:56494', 'dashboard_agent_listen_port': 52365, 'node_id': 'c71e3e56413e8908a8bcd2d19d36749ba3a5602998a4c1a319a3a30e'})
```
- Connect to existing ray cluster with ray client
```
>> import ray
>> ray.init("ray://localhost:25553")
File descriptor limit 2560 is too low for production servers and may result in connection errors. At least 8192 is recommended. --- Fix with 'ulimit -n 8192'
Out[2]: ClientContext(dashboard_url='external_dashboard_url/path1/?query_param1=val1&query_param2=val2', python_version='3.7.13', ray_version='3.0.0.dev0', ray_commit='{{RAY_COMMIT_SHA}}', protocol_version='2022-07-24', _num_clients=1, _context_to_restore=<ray.util.client._ClientContext object at 0x7fb3c560fc90>)
```
- Create and connect to ray cluster in Jupyter
Screen Shot 2022-08-15 at 11 10 22 PM

- Connecting to existing ray cluster in Jupyter with ray client

Screen Shot 2022-08-15 at 11 11 24 PM

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/worker.py
Comment (rickyyx) [inline python/ray/_private/worker.py:911]:
```suggestion
        scheme = f"{parsed_url.scheme}://"
```

**Review by rickyyx — COMMENTED** [review]
Thanks @nikitavemuri for putting this fix quickly! 

The code LGTM, and I only have a few comments on tests if possible. (Doing a manual CLI test and attach them to the PR to test the behaviour of state api and job api when this is set could be an alternative given the short release timeline)   
1.  Do you think it would be possible to add some integration tests for job and state API sanity test when the override exists? For example, you could monkeypatch this env and do a few state queries like [here](https://github.com/ray-project/ray/blob/6084ee5a633c4a9ac100121d24a51a40867cba1f/python/ray/tests/test_state_api.py#L1573). I am not 100% sure what are the best job submission jobs to replicate, but I think [these](https://github.com/ray-project/ray/blob/010a3566e68fd066cecbb88328e720dde3f1b059/dashboard/modules/snapshot/tests/test_job_submission.py#L37) look promising. 
2. I think it will also be great to add a test to `test_ray_init_2` or somewhere we test the external url override with ray init w/o dashboard.

File: dashboard/modules/snapshot/tests/test_job_submission.py
Comment (nikitavemuri) [inline dashboard/modules/snapshot/tests/test_job_submission.py:56]: When an override dashboard url is specified, `ray_context.address_info["webui_url"]` may not be an IP address which can be parsed in `wait_until_server_available`. `ray._private.worker._global_node.webui_url` will contain the local dashboard IP address though. I thought this change was ok because this address is being passed in to the public `JobSubmissionClient` API and is not being formed internally.

File: dashboard/modules/snapshot/tests/test_job_submission.py
Comment (rickyyx) [inline dashboard/modules/snapshot/tests/test_job_submission.py:13]: why is the `noqa` here?

File: dashboard/modules/snapshot/tests/test_job_submission.py
Comment (rickyyx) [inline dashboard/modules/snapshot/tests/test_job_submission.py:50]: Could we actually do `monkeypatch` for the env rather than overriding sys env? Directly modifying ENV in tests created issues before AFAIK. 


For example: https://github.com/ray-project/ray/blob/6084ee5a633c4a9ac100121d24a51a40867cba1f/python/ray/tests/test_state_api.py#L1965-L1971

File: python/ray/tests/test_state_api.py
Comment (rickyyx) [inline python/ray/tests/test_state_api.py:1618]: We don't need to clean up like this if we use `monkeypath` I believe. (but we will need do `ray.init` manually in the test rather than using the `shutdown_only` fixture).

File: python/ray/tests/test_state_api.py
Comment (rickyyx) [inline python/ray/tests/test_state_api.py:2695]: why are we adding special rules for linters here?

**Review by rickyyx — APPROVED** [review]
LGTM other than the suggestion of using `monkeypatch` in the tests, as well as some documentation suggests. 
<div><br/>via <a href="https://giphy.com/gifs/eating-mrw-yyZRSvISN1vvW">GIPHY</a></div>

Will wait for committers' to review and merge.

File: dashboard/modules/snapshot/tests/test_job_submission.py
Comment (rickyyx) [reply dashboard/modules/snapshot/tests/test_job_submission.py:56]: I see, so in general, if we use:
- `ray_context.address_info["webui_url"]`: this is the overriden URL if set. And not guaranteed to be queryable with APIs like in state API/jobs
- `ray._private.worker._global_node.webui_url`: this is the old `<host>:<port>` url, that refers to the address of the dashboard (if present) directly. 

Is this right? If so, would you also documenting this invariance somewhere in the code? Maybe when `ray_context` is first created so users of this would know the subtle differences here.

File: python/ray/tests/test_state_api.py
Comment (rickyyx) [inline python/ray/tests/test_state_api.py:1583]: i think we could duplicate the original test, and rename it to something more representative of what's this doing. e.g. `test_state_api_with_external_dashboard_override`.

File: dashboard/modules/snapshot/tests/test_job_submission.py
Comment (nikitavemuri) [reply dashboard/modules/snapshot/tests/test_job_submission.py:50]: Got it. I'll also update the other tests and fixtures to use monkeypatch

Comment (rickyyx) [reply dashboard/modules/snapshot/tests/test_job_submission.py:50]: thanks, i should have been more explicit in my previous nits.

File: dashboard/modules/snapshot/tests/test_job_submission.py
Comment (nikitavemuri) [reply dashboard/modules/snapshot/tests/test_job_submission.py]: This was originally needed because fixtures don't follow the import lint rules, but I updated to move the fixture to a `conftest.py` file so this lint ignore can be removed

File: python/ray/tests/test_state_api.py
Comment (nikitavemuri) [reply python/ray/tests/test_state_api.py]: This is no longer required because we are using monkeypatch to set the env var

File: python/ray/tests/test_state_api.py
Comment (nikitavemuri) [reply python/ray/tests/test_state_api.py]: Yes,`ray.init` is already called in this test but can we still use `shutdown_only` to clean up the cluster after?

File: python/ray/tests/test_state_api.py
Comment (nikitavemuri) [reply python/ray/tests/test_state_api.py]: When `override_url` is set to `None` this should be the same as the original test. Does it still make sense to duplicate?

Comment (rickyyx) [reply python/ray/tests/test_state_api.py]: I think so - since there are also `override_url` values right?

Comment (nikitavemuri) [reply python/ray/tests/test_state_api.py]: ok, updated

**Review by rkooo567 — APPROVED** [review]
The ideal fix imo is to write overridden URL to internal kv and query from there (in that way, the output will not change whenever the env var is changed). But if this env var is not exposed as public, I am okay with the current change.

**Comment by alanwguo** [comment]
> The ideal fix imo is to write overridden URL to internal kv and query from there (in that way, the output will not change whenever the env var is changed). But if this env var is not exposed as public, I am okay with the current change.

what do you mean by that? Isn't the problem that the overriden url is not something that should be used by internal implementation?

There seems to be two separate concepts here:

Dashboard UI URL
Dashboard API base address / hostname

Right now the code base assumes dashboard_url in internal_kv is the second concept right?

**Review by alanwguo — APPROVED** [review]

**Pull request merged**
